pkg/condition/condition.go

```diff
@@ -0,0 +1,149 @@
+package condition
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"os"
+	"strings"
+
+	operators "github.com/operator-framework/api/pkg/operators/v2"
+	"github.com/pkg/errors"
+	meta "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	patcher "github.com/openshift/windows-machine-config-operator/pkg/patch"
+	"github.com/openshift/windows-machine-config-operator/pkg/retry"
+)
+
+const (
+	upgradeableTrueReason       = "upgradeIsSafe"
+	upgradeableTrueMessage      = "The operator is safe for upgrade"
+	upgradeableFalseReason      = "upgradeIsNotSafe"
+	upgradeableFalseMessageBase = "The operator is currently processing sub-components"
+	// operatorConditionName is an environment variable set by OLM identifying the operator's OperatorCondition CR
+	operatorConditionName = "OPERATOR_CONDITION_NAME"
+)
+
+// OcName is the name of the OLM-managed OperatorCondition resource. Empty if operator is not OLM-managed
+var OcName string
+
+// init runs once, initializing global variables
+func init() {
+	OcName = getResourceName()
+}
+
+// EnableOLMUpgrade is a wrapper that sets the Upgradeable condition to True to allow OLM to perform operator upgrades
+func EnableOLMUpgrade(c client.Client, watchNamespace, controllerName string) error {
+	return set(c, watchNamespace, controllerName, operators.Upgradeable, meta.ConditionTrue,
+		upgradeableTrueReason, upgradeableTrueMessage)
+}
+
+// DisableOLMUpgrade is a wrapper that sets the Upgradeable condition to False to prevent operator upgrades by OLM
+func DisableOLMUpgrade(c client.Client, watchNamespace, controllerName string) error {
+	return set(c, watchNamespace, controllerName, operators.Upgradeable, meta.ConditionFalse,
+		upgradeableFalseReason, fmt.Sprintf("%s: %s", upgradeableFalseMessageBase, controllerName))
+}
+
+// set sets the given condition's values within the associated OperatorCondition.
+// Only returns after waiting for a made change to take effect. No-op if operator is not OLM-managed
+func set(c client.Client, ocNamespace, controllerName, conditionType string, status meta.ConditionStatus,
+	reason, message string) error {
+	// Check if operator is OLM-managed
+	if OcName == "" {
+		return nil
+	}
+
+	oc, err := get(c, ocNamespace)
+	if err != nil {
+		return err
+	}
+	// Do not change the Upgradeable condition if a different controller is busy. Prevents potential timing errors
+	// if controller A, which started after controller B, finishes reconciling before controller B finishes
```

Comment:
Do we need a lock?

Reply:
I don't think a lock is needed. With a lock, if 2 controllers are reconciling at the same time, a go routine from one controller would be stuck waiting for the lock in order to proceed despite the Upgradeable condition already reflecting the proper operator state.